src/portable/espressif/esp32s2/dcd_esp32s2.c

```diff
@@ -0,0 +1,749 @@
+                                 desc_edpt->wMaxPacketSize.size << USB_MPS0_S;
+        USB0.daintmsk |= (1 << (16 + epnum));
+    } else {
+        // Peripheral FIFO architecture (Rev18 RM 29.11)
+        //
+        // --------------- 320 ( 1280 bytes )
+        // | IN FIFO 3  |
+        // --------------- y + x + 16 + GRXFSIZ
+        // | IN FIFO 2  |
+        // --------------- x + 16 + GRXFSIZ
+        // | IN FIFO 1  |
+        // --------------- 16 + GRXFSIZ
+        // | IN FIFO 0  |
+        // --------------- GRXFSIZ
+        // | OUT FIFO   |
+        // | ( Shared ) |
+        // --------------- 0
+        //
+        // Since OUT FIFO = 50, FIFO 0 = 16, average of FIFOx = (312-50-16) / 3 = 82 ~ 80
+        in_ep[epnum].diepctl |= USB_D_USBACTEP1_M |
+                                (epnum - 1) << USB_D_TXFNUM1_S |
+                                desc_edpt->bmAttributes.xfer << USB_D_EPTYPE1_S |
+                                (desc_edpt->bmAttributes.xfer != TUSB_XFER_ISOCHRONOUS ? (1 << USB_DI_SETD0PID1_S) : 0) |
+                                desc_edpt->wMaxPacketSize.size << 0;
+        USB0.daintmsk |= (1 << (0 + epnum));
+
+        // Both TXFD and TXSA are in unit of 32-bit words
+        uint16_t const fifo_size = 80;
+        uint32_t const fifo_offset = (USB0.grxfsiz & USB_NPTXFDEP_V) + 16 + fifo_size * (epnum - 1);
+        USB0.dieptxf[epnum - 1] = (80 << USB_NPTXFDEP_S) | fifo_offset;
+    }
+    return true;
+}
+
+bool dcd_edpt_xfer(uint8_t rhport, uint8_t ep_addr, uint8_t *buffer, uint16_t total_bytes)
+{
+
+    (void)rhport;
+
+    uint8_t const epnum = tu_edpt_number(ep_addr);
+    uint8_t const dir = tu_edpt_dir(ep_addr);
+
+    xfer_ctl_t *xfer = XFER_CTL_BASE(epnum, dir);
+    xfer->buffer = buffer;
+    xfer->total_len = total_bytes;
+    xfer->queued_len = 0;
+    xfer->short_packet = false;
+
+    uint16_t num_packets = (total_bytes / xfer->max_size);
+    uint8_t short_packet_size = total_bytes % xfer->max_size;
+
+    // Zero-size packet is special case.
+    if (short_packet_size > 0 || (total_bytes == 0)) {
+        num_packets++;
+    }
+
+    ESP_LOGV(TAG, "Transfer <-> EP%i, %s, pkgs: %i, bytes: %i",
+             epnum, ((dir == TUSB_DIR_IN) ? "USB0.HOST (in)" : "HOST->DEV (out)"),
+             num_packets, total_bytes);
+
+    // IN and OUT endpoint xfers are interrupt-driven, we just schedule them
+    // here.
+    if (dir == TUSB_DIR_IN) {
+        // A full IN transfer (multiple packets, possibly) triggers XFRC.
+        int bytes2fifo_left = total_bytes;
+        uint32_t val; // 32 bit val from 4 buff addresses
+
+        USB0.in_ep_reg[epnum].diepint = ~0U; // clear all ints
+        USB0.in_ep_reg[epnum].dieptsiz = (num_packets << USB_D_PKTCNT0_S) | total_bytes;
+        USB0.in_ep_reg[epnum].diepctl |= USB_D_EPENA1_M | USB_D_CNAK1_M; // Enable | CNAK
+        while (bytes2fifo_left > 0) {  // TODO move it to ep_in_handle (IDF-1475)
```

Comment:
@cr1901 this seems to be an issue, the setup you pointed out shouldn't cause the overrun endpoint size.

Reply:
Indeed, I wasn't sure whether the setup problem could cause overrun or not, but it should still be fixed.

I don't think `CFG_TUD_ENDOINT0_SIZE` can be < 8 per spec. 

In my synopsys port, it looks like I have [a guard](https://github.com/hathach/tinyusb/blob/master/src/portable/st/synopsys/dcd_synopsys.c#L530-L538) for this. It's also in [this file](https://github.com/hathach/tinyusb/pull/319/files/748cc88769ffd2039fcb0f4db9bb2cb4bc84965f#diff-371e8160b188a80c1a3821c75d194adfR522-R528) as well.

Instead of doing an initial xfer in `dcd_edpt_xfer`, I opt to instead deliberately [trigger](https://github.com/hathach/tinyusb/blob/master/src/portable/st/synopsys/dcd_synopsys.c#L356) an IN endpoint ISR to start the transfer. This avoids me having to write out that code twice :P.

Is there any particular constraint on this version of the Synopsys core that makes kickstarting the ISR impossible?

Reply:
@cr1901 it is comment out on purpose from ESP, there is possibly some configuration difference comparing to stm32. I tried to enable the use of  `dtknqr4_fifoemptymsk` ( ESP's name for DIEPEMPMSK) in both edpt_xfer and handle_epin_ints but it results in no data transfer at all --> failed to get descriptor. Still doing trials and errors :sweat_smile: :sweat_smile: